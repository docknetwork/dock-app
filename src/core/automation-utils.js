export function addTestId(testId) {
  return {
    testID: testId,
    accessibilityLabel: testId,
  };
}
