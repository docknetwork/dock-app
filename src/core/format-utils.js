export function formatCurrency(value, currency = 'USD') {
  var formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,

    minimumFractionDigits: 2,
    maximumFractionDigits: 10,
  });

  return formatter.format(value);
}

export function formatAddress(value, size = 19) {
  if (!value || size > value.length) {
    return value;
  }

  const offset = size / 2;
  return `${value.substring(0, offset)}...${value.substring(
    value.length - offset,
  )}`;
}
