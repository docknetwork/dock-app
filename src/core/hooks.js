import {useEffect} from 'react';
import {BackHandler} from 'react-native';

export function useBackHandler(callback) {
  useEffect(() => {
    const backHandler = BackHandler.addEventListener(
      'hardwareBackPress',
      () => {
        console.log('stuff');

        return true;
      },
    );

    return () => backHandler.remove();
  }, []);
}
