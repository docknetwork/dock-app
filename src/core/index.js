/**
 * JSON.stringify with error handler
 *
 * @param value
 * @returns {string}
 */
export function getJsonOrError(value) {
  try {
    return JSON.stringify(value);
  } catch (err) {
    return err;
  }
}
