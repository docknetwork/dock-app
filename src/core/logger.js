const winston = require('winston');

export const Logger = winston.createLogger({
  format: winston.format.simple(),
  levels: ['debug'],
  transports: [new winston.transports.Console()],
});
