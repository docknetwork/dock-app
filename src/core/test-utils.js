import {NativeBaseProvider} from 'native-base';
import React from 'react';

const inset = {
  frame: {x: 0, y: 0, width: 0, height: 0},
  insets: {top: 0, left: 0, right: 0, bottom: 0},
};

export const renderAppProviders = component => {
  return (
    <NativeBaseProvider initialWindowMetrics={inset}>
      {component}
    </NativeBaseProvider>
  );
};

export const mockState = (rootKey, state) => {
  return () => ({
    [rootKey]: state,
  });
};

export const mockDispatch = result => () => result;
