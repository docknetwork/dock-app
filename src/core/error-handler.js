import { Stack } from 'native-base';
import React from 'react';
import { Typography } from 'src/design-system';
import { translate } from 'src/locales';

export class ErrorBoundary extends React.Component<any, any> {
   constructor(props: any) {
      super(props);
      this.state = { hasError: false };
   }

   static getDerivedStateFromError(error: any) {
      return { hasError: true };
   }

   componentDidCatch(error: any, errorInfo: any) {
      // TODO: log the error in sentry
      console.error(error);
   }

   render() {
      if (this.state.hasError) {
         return <Stack><Typography>{translate('global.rendering_error')}</Typography></Stack>;
      }

      return this.props.children;
   }
}

export const withErrorBoundary = (WrapperComponent: any) => (props: any) => {
   return (
      <ErrorBoundary>
         <WrapperComponent {...props} />
      </ErrorBoundary>
   );
};
