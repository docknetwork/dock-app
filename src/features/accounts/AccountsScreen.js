import {Menu, Pressable, ScrollView, Stack} from 'native-base';
import React, {useCallback, useEffect, useState} from 'react';
import {Platform, RefreshControl} from 'react-native';
import DocumentPicker from 'react-native-document-picker';
import RNExitApp from 'react-native-exit-app';
import RNFS from 'react-native-fs';
import {useDispatch, useSelector} from 'react-redux';
import {addTestId} from 'src/core/automation-utils';
import {withErrorBoundary} from 'src/core/error-handler';
import {formatCurrency} from 'src/core/format-utils';
import {navigate} from 'src/core/navigation';
import {Routes} from 'src/core/routes';
import {translate} from 'src/locales';
import CogIcon from '../../assets/icons/cog.svg';
import DocumentDownloadIcon from '../../assets/icons/document-download.svg';
import PlusCircleWhiteIcon from '../../assets/icons/plus-circle-white.svg';
import PlusCircleIcon from '../../assets/icons/plus-circle.svg';
import {PolkadotIcon} from '../../components/PolkadotIcon';
import {
  AlertIcon,
  BigButton,
  Box,
  Button,
  ChevronRightIcon,
  DotsVerticalIcon,
  Footer,
  Header,
  IconButton,
  NBox,
  ScreenContainer,
  Theme,
  Typography,
} from '../../design-system';
import {createAccountOperations} from '../account-creation/create-account-slice';
import {TokenAmount} from '../tokens/ConfirmTransactionModal';
import {accountOperations, accountSelectors} from './account-slice';
import {AddAccountModal} from './AddAccountModal';
import {AccountsScreenTestIDs} from './test-ids';

export const AccountsScreen = withErrorBoundary(
  ({
    accounts = [],
    onAddAccount,
    onImportExistingAccount,
    onDelete,
    onEdit,
    onDetails,
    onSettings,
    onRefresh,
    isRefreshing,
  }) => {
    const isEmpty = accounts.length === 0;
    const [showAddAccount, setShowAddAccount] = useState();
    const [showImportAccount, setShowImportAccount] = useState();

    return (
      <ScreenContainer testID={AccountsScreenTestIDs.screen}>
        <Header>
          <Box
            marginLeft={22}
            marginRight={22}
            flexDirection="row"
            alignItems="center">
            <Box flex>
              <Typography
                fontFamily="Montserrat"
                fontSize={24}
                fontWeight="600">
                {translate('account_list.title')}
              </Typography>
            </Box>
            <Box row>
              <IconButton
                col
                {...addTestId(AccountsScreenTestIDs.addAccountMenuBtn)}
                marginRight={10}
                onPress={() => setShowAddAccount(true)}>
                <PlusCircleWhiteIcon />
              </IconButton>
              <IconButton col onPress={onSettings}>
                <CogIcon />
              </IconButton>
            </Box>
          </Box>
        </Header>
        <ScrollView
          refreshControl={
            <RefreshControl refreshing={isRefreshing} onRefresh={onRefresh} />
          }>
          <Stack mx={26} flex={1}>
            {isEmpty ? (
              <Box flex justifyContent="center" alignItems="center">
                <Typography marginTop={12}>
                  {translate('account_list.empty_accounts')}
                </Typography>
              </Box>
            ) : (
              <NBox>
                {accounts.map(account => {
                  return (
                    <Stack
                      direction="row"
                      borderRadius={12}
                      backgroundColor={Theme.colors.secondaryBackground}
                      space={2}
                      mb={4}
                      py={6}
                      px={6}>
                      <Stack direction="column" flex={1}>
                        <Stack direction="row" alignItems="center">
                          <Pressable
                            _pressed={{
                              opacity: Theme.touchOpacity,
                            }}
                            onPress={() => onDetails(account)}
                            flex={1}>
                            <Stack direction="row" flex={1} alignItems="center">
                              <PolkadotIcon address={account.id} size={32} />
                              <Stack direction="row" flex={1} ml={3}>
                                <Typography
                                  color={Theme.colors.textHighlighted}
                                  fontWeight={600}>
                                  {account.name}
                                </Typography>
                                <ChevronRightIcon marginTop={1} />
                              </Stack>
                            </Stack>
                          </Pressable>
                          <NBox py={1} px={1}>
                            <Stack direction="row">
                              {!account.hasBackup ||
                              (account.meta &&
                                account.meta.keypairNotFoundWarning) ? (
                                <NBox mr={3} mt={1}>
                                  <AlertIcon />
                                </NBox>
                              ) : null}
                              <Menu
                                trigger={triggerProps => {
                                  return (
                                    <Pressable
                                      {...triggerProps}
                                      _pressed={{
                                        opacity: Theme.touchOpacity,
                                      }}>
                                      <DotsVerticalIcon />
                                    </Pressable>
                                  );
                                }}>
                                <Menu.Item onPress={() => onDetails(account)}>
                                  {translate('account_list.account_details')}
                                </Menu.Item>
                                <Menu.Item onPress={() => onDelete(account)}>
                                  {translate('account_list.delete_account')}
                                </Menu.Item>
                              </Menu>
                            </Stack>
                          </NBox>
                        </Stack>

                        <TokenAmount amount={account.balance}>
                          {({
                            fiatAmount,
                            fiatSymbol,
                            tokenAmount,
                            tokenSymbol,
                          }) => (
                            <>
                              <Stack direction="column" mt={4}>
                                <Typography variant="h2">
                                  {tokenAmount} {tokenSymbol}
                                </Typography>
                                <Typography fontSize="14px">
                                  {formatCurrency(fiatAmount)}
                                </Typography>
                              </Stack>
                            </>
                          )}
                        </TokenAmount>

                        <Stack direction="row" mt={4}>
                          {
                            <Button
                              width="50%"
                              size="sm"
                              disabled={account.readOnly}
                              colorScheme="dark"
                              {...addTestId('tokenSend')}
                              onPress={() => {
                                navigate(Routes.TOKEN_SEND, {
                                  address: account.id,
                                });
                              }}>
                              {translate('account_list.send_token')}
                            </Button>
                          }
                          <Button
                            width="50%"
                            size="sm"
                            ml={2}
                            colorScheme="dark"
                            {...addTestId('tokenReceive')}
                            onPress={() => {
                              navigate(Routes.TOKEN_RECEIVE, {
                                address: account.id,
                              });
                            }}>
                            {translate('account_list.receive_token')}
                          </Button>
                        </Stack>
                      </Stack>
                    </Stack>
                  );
                })}
              </NBox>
            )}
          </Stack>
        </ScrollView>
        {isEmpty ? (
          <Footer marginBottom={114} marginLeft={26} marginRight={26} flex>
            <BigButton
              {...addTestId('CreateNewAccount')}
              onPress={onAddAccount}
              icon={<PlusCircleIcon />}>
              Create new account
            </BigButton>
            <BigButton
              {...addTestId('add_account_modal.import_existing')}
              onPress={() => {
                setShowImportAccount(true);
                setShowAddAccount(true);
              }}
              icon={<DocumentDownloadIcon />}>
              {translate('add_account_modal.import_existing')}
            </BigButton>
          </Footer>
        ) : null}
        <AddAccountModal
          visible={showAddAccount}
          showImportAccount={showImportAccount}
          onClose={() => {
            setShowImportAccount(false);
            setShowAddAccount(false);
          }}
          onAddAccount={onAddAccount}
          onImportExistingAccount={onImportExistingAccount}
        />
      </ScreenContainer>
    );
  },
);

export const AccountsContainer = withErrorBoundary(({navigation}) => {
  const dispatch = useDispatch();
  const accounts = useSelector(accountSelectors.getAccounts);
  const [isRefreshing, setRefreshing] = useState(false);

  const onRefresh = useCallback(() => {
    setRefreshing(true);
    dispatch(accountOperations.fetchBalances()).finally(() => {
      setRefreshing(false);
    });
  }, [dispatch]);

  useEffect(() => {
    dispatch(accountOperations.loadAccounts());
  }, [dispatch]);

  useEffect(() => {
    if (Platform.OS === 'android') {
      navigation.addListener('beforeRemove', e => {
        RNExitApp.exitApp();
        e.preventDefault();
      });
    }
  }, [navigation]);

  return (
    <AccountsScreen
      onDelete={accountId => {
        dispatch(accountOperations.removeAccount(accountId));
      }}
      onDetails={account => {
        navigate(Routes.ACCOUNT_DETAILS, {
          id: account.id,
        });
      }}
      isRefreshing={isRefreshing}
      onRefresh={onRefresh}
      accounts={accounts}
      onAddAccount={() => {
        dispatch(accountOperations.addAccountFlow());
      }}
      onSettings={() => {
        navigate(Routes.APP_SETTINGS);
      }}
      onImportExistingAccount={async method => {
        if (method === 'mnemonic') {
          navigate(Routes.ACCOUNT_IMPORT_FROM_MNEMONIC);
        } else if (method === 'qrcode') {
          navigate(Routes.APP_QR_SCANNER, {
            onData: data => {
              dispatch(createAccountOperations.importFromJson(data));
            },
          });
        } else if (method === 'json') {
          const files = await DocumentPicker.pick({
            type: [DocumentPicker.types.allFiles],
          });

          const fileData = await RNFS.readFile(files[0].fileCopyUri);

          dispatch(createAccountOperations.importFromJson(fileData));
        }
      }}
    />
  );
});
