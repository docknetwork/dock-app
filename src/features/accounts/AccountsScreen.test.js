import React from 'react';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import {AccountsScreen} from './AccountsScreen';
import {renderAppProviders} from '../../core/test-utils';

const mockAccounts = [
  {
    id: Date.now(),
    name: 'Account name',
    balance: 21,
  },
];

describe('AccountsScreen', () => {
  it('should render correctly with no accounts', () => {
    const tree = shallow(renderAppProviders(<AccountsScreen accounts={[]} />));
    expect(tree.dive()).toMatchSnapshot();
  });

  it('should render correctly with accounts', () => {
    const tree = shallow(
      renderAppProviders(<AccountsScreen accounts={mockAccounts} />),
    );
    expect(tree.dive()).toMatchSnapshot();
  });
});
