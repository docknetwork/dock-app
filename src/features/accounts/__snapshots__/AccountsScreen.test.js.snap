// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsScreen should render correctly with accounts 1`] = `
<NativeBaseProvider
  initialWindowMetrics={
    Object {
      "frame": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "insets": Object {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
    }
  }
>
  <Component
    accounts={
      Array [
        Object {
          "balance": 21,
          "id": "test",
          "name": "Account name",
        },
      ]
    }
  />
</NativeBaseProvider>
`;

exports[`AccountsScreen should render correctly with no accounts 1`] = `
<NativeBaseProvider
  initialWindowMetrics={
    Object {
      "frame": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "insets": Object {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      },
    }
  }
>
  <Component
    accounts={Array []}
  />
</NativeBaseProvider>
`;
